File: go/sketchkit/bloom/bloom.go
// SPDX-License-Identifier: Apache-2.0
// Code authors: Vijay Erramilli and Codex
package bloom

import (
	"errors"
	"fmt"
	"math"
	"math/bits"

	sketchhash "github.com/llm-measurement/llm-sketchkit/go/sketchkit/hash"
	"github.com/llm-measurement/llm-sketchkit/go/sketchkit/internal/hashfamily"
	sketchpb "github.com/llm-measurement/llm-sketchkit/go/sketchkit/internal/pb"
	"google.golang.org/protobuf/proto"
)

const (
	wireVersion  = 1
	maxWireBytes = 4 << 20
)

// Profile names a Bloom profile from spec/profiles.md.
type Profile string

const (
	ProfileMicro   Profile = "micro"
	ProfileSmall   Profile = "small"
	ProfileDefault Profile = "default"
)

var (
	// ErrUnknownProfile reports a profile outside spec/profiles.md.
	ErrUnknownProfile = errors.New("unknown bloom profile")

	// ErrInvalidShape reports a non-profile Bloom bit/hash-count pair.
	ErrInvalidShape = errors.New("invalid bloom shape")

	// ErrIncompatibleMerge reports metadata mismatch on merge.
	ErrIncompatibleMerge = errors.New("incompatible bloom merge")

	// ErrCountOverflow reports inserted-count overflow.
	ErrCountOverflow = errors.New("bloom inserted count overflow")

	// ErrInvalidWireEncoding reports a malformed serialized sketch.
	ErrInvalidWireEncoding = errors.New("invalid bloom wire encoding")
)

type profileConfig struct {
	ratedInsertions uint64
	targetFPR       float64
	bitCount        uint64
	hashCount       uint32
}

var profileConfigs = map[Profile]profileConfig{
	ProfileMicro:   {ratedInsertions: 10_000, targetFPR: 0.001, bitCount: 143_776, hashCount: 10},
	ProfileSmall:   {ratedInsertions: 100_000, targetFPR: 0.001, bitCount: 1_437_759, hashCount: 10},
	ProfileDefault: {ratedInsertions: 1_000_000, targetFPR: 0.0001, bitCount: 19_170_117, hashCount: 13},
}

// Sketch is a mergeable Bloom filter over pre-hashed uint64 values.
type Sketch struct {
	profile   Profile
	config    profileConfig
	domain    sketchhash.Domain
	algorithm sketchhash.Algorithm

	bitset        []byte
	insertedCount uint64
}

// New constructs an empty Bloom sketch from a named profile.
func New(profile Profile, domain sketchhash.Domain, algorithm sketchhash.Algorithm) (*Sketch, error) {
	config, ok := profileConfigs[profile]
	if !ok {
		return nil, fmt.Errorf("%w: %s", ErrUnknownProfile, profile)
	}
	if !sketchhash.IsRegisteredDomain(domain) {
		return nil, fmt.Errorf("%w: %s", sketchhash.ErrUnregisteredDomain, domain)
	}
	if algorithm != sketchhash.HMACSHA25664 {
		return nil, fmt.Errorf("%w: %s", ErrIncompatibleMerge, algorithm)
	}

	return newSketch(profile, config, domain, algorithm), nil
}

func newSketch(
	profile Profile,
	config profileConfig,
	domain sketchhash.Domain,
	algorithm sketchhash.Algorithm,
) *Sketch {
	return &Sketch{
		profile:   profile,
		config:    config,
		domain:    domain,
		algorithm: algorithm,
		bitset:    make([]byte, byteLen(config.bitCount)),
	}
}

// AddHash inserts a pre-hashed value.
func (s *Sketch) AddHash(hash uint64) error {
	if s.insertedCount == math.MaxUint64 {
		return ErrCountOverflow
	}
	for i := uint32(0); i < s.config.hashCount; i++ {
		s.setBit(s.location(hash, i))
	}
	s.insertedCount++

	return nil
}

// MayContainHash returns true when hash may have been inserted.
func (s *Sketch) MayContainHash(hash uint64) bool {
	for i := uint32(0); i < s.config.hashCount; i++ {
		if !s.bit(s.location(hash, i)) {
			return false
		}
	}

	return true
}

// Merge ORs other into s. Metadata must match under the v0.1 merge policy.
func (s *Sketch) Merge(other *Sketch) error {
	if other == nil {
		return nil
	}
	if s.profile != other.profile ||
		s.config.bitCount != other.config.bitCount ||
		s.config.hashCount != other.config.hashCount ||
		s.domain != other.domain ||
		s.algorithm != other.algorithm {
		return fmt.Errorf("%w: metadata differs", ErrIncompatibleMerge)
	}
	if s.insertedCount > math.MaxUint64-other.insertedCount {
		return ErrCountOverflow
	}
	for i := range s.bitset {
		s.bitset[i] |= other.bitset[i]
	}
	s.insertedCount += other.insertedCount

	return nil
}

// FalsePositiveEstimate returns the Bloom FPR estimate from the serialized
// bitset fill ratio. This keeps estimates merge-exact after OR merges.
func (s *Sketch) FalsePositiveEstimate() float64 {
	fillRatio := float64(s.SetBitCount()) / float64(s.config.bitCount)
	return math.Pow(fillRatio, float64(s.config.hashCount))
}

// InsertedCount returns the number of AddHash calls observed by this sketch.
func (s *Sketch) InsertedCount() uint64 {
	return s.insertedCount
}

// BitCount returns the configured bit count.
func (s *Sketch) BitCount() uint64 {
	return s.config.bitCount
}

// SetBitCount returns the number of one bits in the serialized bitset.
func (s *Sketch) SetBitCount() uint64 {
	var count uint64
	for _, b := range s.bitset {
		count += uint64(bits.OnesCount8(b))
	}

	return count
}

// HashCount returns the configured hash count.
func (s *Sketch) HashCount() uint32 {
	return s.config.hashCount
}

// RatedInsertions returns the profile's rated insertion count.
func (s *Sketch) RatedInsertions() uint64 {
	return s.config.ratedInsertions
}

// TargetFPR returns the profile's target false-positive rate.
func (s *Sketch) TargetFPR() float64 {
	return s.config.targetFPR
}

// RepresentationMode returns the current protobuf representation mode.
func (s *Sketch) RepresentationMode() sketchpb.RepresentationMode {
	return sketchpb.RepresentationMode_REPRESENTATION_MODE_BLOOM_BITSET
}

// Clone returns a deep copy of s.
func (s *Sketch) Clone() *Sketch {
	clone := newSketch(s.profile, s.config, s.domain, s.algorithm)
	clone.insertedCount = s.insertedCount
	copy(clone.bitset, s.bitset)

	return clone
}

// MarshalBinary returns deterministic protobuf bytes.
func (s *Sketch) MarshalBinary() ([]byte, error) {
	message := s.toProto()
	return proto.MarshalOptions{Deterministic: true}.Marshal(message)
}

// Parse decodes a Bloom sketch from deterministic protobuf bytes.
func Parse(data []byte) (*Sketch, error) {
	if len(data) > maxWireBytes {
		return nil, fmt.Errorf("%w: input length %d exceeds %d", ErrInvalidWireEncoding, len(data), maxWireBytes)
	}

	var message sketchpb.Sketch
	if err := proto.Unmarshal(data, &message); err != nil {
		return nil, err
	}

	return fromProto(&message)
}

func (s *Sketch) location(hash uint64, index uint32) uint64 {
	return hashfamily.BloomHash(hash, index) % s.config.bitCount
}

func (s *Sketch) setBit(position uint64) {
	s.bitset[position/8] |= byte(1 << (position % 8))
}

func (s *Sketch) bit(position uint64) bool {
	return s.bitset[position/8]&byte(1<<(position%8)) != 0
}

func (s *Sketch) toProto() *sketchpb.Sketch {
	bitCount := s.config.bitCount
	hashCount := s.config.hashCount
	return &sketchpb.Sketch{
		Metadata: &sketchpb.SketchMetadata{
			Kind:               sketchpb.SketchKind_SKETCH_KIND_BLOOM,
			WireVersion:        wireVersion,
			Profile:            string(s.profile),
			HashDomain:         string(s.domain),
			HashAlgo:           sketchpb.HashAlgorithm_HASH_ALGORITHM_HMAC_SHA256_64,
			RepresentationMode: s.RepresentationMode(),
			BloomBitCount:      &bitCount,
			BloomHashCount:     &hashCount,
		},
		Body: &sketchpb.Sketch_Bloom{
			Bloom: &sketchpb.BloomSketch{
				Bitset:        append([]byte(nil), s.bitset...),
				InsertedCount: s.insertedCount,
			},
		},
	}
}

func fromProto(message *sketchpb.Sketch) (*Sketch, error) {
	metadata := message.GetMetadata()
	body := message.GetBloom()
	if metadata == nil || body == nil {
		return nil, fmt.Errorf("%w: missing bloom metadata or body", ErrInvalidWireEncoding)
	}
	if metadata.GetKind() != sketchpb.SketchKind_SKETCH_KIND_BLOOM {
		return nil, fmt.Errorf("%w: kind %s", ErrInvalidWireEncoding, metadata.GetKind())
	}
	if metadata.GetWireVersion() != wireVersion {
		return nil, fmt.Errorf("%w: wire version %d", ErrInvalidWireEncoding, metadata.GetWireVersion())
	}
	if metadata.GetHashAlgo() != sketchpb.HashAlgorithm_HASH_ALGORITHM_HMAC_SHA256_64 {
		return nil, fmt.Errorf("%w: hash algorithm %s", ErrInvalidWireEncoding, metadata.GetHashAlgo())
	}
	if metadata.GetRepresentationMode() != sketchpb.RepresentationMode_REPRESENTATION_MODE_BLOOM_BITSET {
		return nil, fmt.Errorf("%w: representation %s", ErrInvalidWireEncoding, metadata.GetRepresentationMode())
	}

	profile := Profile(metadata.GetProfile())
	config, ok := profileConfigs[profile]
	if !ok {
		return nil, fmt.Errorf("%w: %s", ErrUnknownProfile, profile)
	}
	if metadata.GetBloomBitCount() != config.bitCount || metadata.GetBloomHashCount() != config.hashCount {
		return nil, fmt.Errorf("%w: profile %s bit_count=%d hash_count=%d",
			ErrInvalidShape, profile, metadata.GetBloomBitCount(), metadata.GetBloomHashCount())
	}
	if !sketchhash.IsRegisteredDomain(sketchhash.Domain(metadata.GetHashDomain())) {
		return nil, fmt.Errorf("%w: %s", sketchhash.ErrUnregisteredDomain, metadata.GetHashDomain())
	}
	if len(body.GetBitset()) != int(byteLen(config.bitCount)) {
		return nil, fmt.Errorf("%w: bitset length", ErrInvalidWireEncoding)
	}
	if !finalBytePaddingZero(body.GetBitset(), config.bitCount) {
		return nil, fmt.Errorf("%w: final byte padding bits set", ErrInvalidWireEncoding)
	}

	sketch := newSketch(profile, config, sketchhash.Domain(metadata.GetHashDomain()), sketchhash.HMACSHA25664)
	sketch.insertedCount = body.GetInsertedCount()
	copy(sketch.bitset, body.GetBitset())

	return sketch, nil
}

func byteLen(bitCount uint64) uint64 {
	return (bitCount + 7) / 8
}

func finalBytePaddingZero(bitset []byte, bitCount uint64) bool {
	remainder := bitCount % 8
	if remainder == 0 || len(bitset) == 0 {
		return true
	}
	mask := byte(0xff << remainder)
	return bitset[len(bitset)-1]&mask == 0
}

File: python/llm_sketchkit/bloom.py
# SPDX-License-Identifier: Apache-2.0
# Code authors: Vijay Erramilli and Codex
"""Pure-Python Bloom sketch implementation."""

from __future__ import annotations

import math
from typing import Any, cast

from . import _proto, hashfamily, profiles

MAX_UINT64 = (1 << 64) - 1


class BloomError(ValueError):
    """Base class for Bloom failures."""


class UnknownProfileError(BloomError):
    """Raised when a profile is not supported."""


class InvalidShapeError(BloomError):
    """Raised when wire metadata uses a non-profile shape."""


class IncompatibleMergeError(BloomError):
    """Raised when merge metadata differs."""


class CountOverflowError(BloomError):
    """Raised when inserted count exceeds uint64."""


class InvalidWireEncodingError(BloomError):
    """Raised when serialized bytes are malformed."""


class Sketch:
    """Mergeable Bloom filter over pre-hashed unsigned 64-bit values."""

    def __init__(
        self,
        profile: str,
        domain: str = profiles.PROMPT_V1,
        algorithm: str = profiles.HMAC_SHA256_64,
    ) -> None:
        config = profiles.BLOOM_PROFILES.get(profile)
        if config is None:
            raise UnknownProfileError(profile)
        if domain not in profiles.REGISTERED_DOMAINS:
            raise IncompatibleMergeError(domain)
        if algorithm != profiles.HMAC_SHA256_64:
            raise IncompatibleMergeError(algorithm)

        self._profile = profile
        self._config = config
        self._domain = domain
        self._algorithm = algorithm
        self._bitset = bytearray(_byte_len(config.bit_count))
        self._inserted_count = 0

    @classmethod
    def parse(cls, data: bytes) -> Sketch:
        """Decode a Bloom sketch from deterministic protobuf bytes."""

        if len(data) > _proto.MAX_WIRE_BYTES:
            raise InvalidWireEncodingError("wire input too large")
        message = _proto.parse_sketch(data)
        metadata = message.metadata
        if not message.HasField("metadata") or not message.HasField("bloom"):
            raise InvalidWireEncodingError("missing Bloom metadata or body")
        if metadata.kind != _proto.SKETCH_KIND_BLOOM:
            raise InvalidWireEncodingError("wrong sketch kind")
        if metadata.wire_version != _proto.WIRE_VERSION:
            raise InvalidWireEncodingError("wrong wire version")
        if metadata.hash_algo != _proto.HASH_ALGORITHM_HMAC_SHA256_64:
            raise InvalidWireEncodingError("wrong hash algorithm")
        if metadata.representation_mode != _proto.REPRESENTATION_BLOOM_BITSET:
            raise InvalidWireEncodingError("wrong representation")

        profile = cast(str, metadata.profile)
        config = profiles.BLOOM_PROFILES.get(profile)
        if config is None:
            raise UnknownProfileError(profile)
        if (
            cast(int, metadata.bloom_bit_count) != config.bit_count
            or cast(int, metadata.bloom_hash_count) != config.hash_count
        ):
            raise InvalidShapeError(profile)
        if cast(str, metadata.hash_domain) not in profiles.REGISTERED_DOMAINS:
            raise IncompatibleMergeError("unregistered domain")

        bitset = cast(bytes, message.bloom.bitset)
        if len(bitset) != _byte_len(config.bit_count):
            raise InvalidWireEncodingError("wrong bitset length")
        if not _final_byte_padding_zero(bitset, config.bit_count):
            raise InvalidWireEncodingError("final byte padding bits set")

        sketch = cls(profile, cast(str, metadata.hash_domain), profiles.HMAC_SHA256_64)
        sketch._bitset = bytearray(bitset)
        sketch._inserted_count = cast(int, message.bloom.inserted_count)
        return sketch

    def add_hash(self, value: int) -> None:
        """Insert a pre-hashed unsigned 64-bit value.

        CountOverflowError leaves self unchanged. Runtime failures and inputs
        outside the documented types have no rollback guarantee.
        """

        if self._inserted_count == MAX_UINT64:
            raise CountOverflowError("inserted count")
        value &= hashfamily.MASK64
        for index in range(self._config.hash_count):
            self._set_bit(self._location(value, index))
        self._inserted_count += 1

    def may_contain_hash(self, value: int) -> bool:
        """Return whether the value may have been inserted."""

        value &= hashfamily.MASK64
        for index in range(self._config.hash_count):
            if not self._bit(self._location(value, index)):
                return False
        return True

    def merge(self, other: Sketch | None) -> None:
        """OR another Bloom sketch with identical metadata into this sketch.

        IncompatibleMergeError and CountOverflowError leave self unchanged.
        None is a no-op; a distinct source is never modified. Runtime failures,
        invalid private state, and concurrent use have no rollback guarantee.
        """

        if other is None:
            return
        if (
            self._profile != other._profile
            or self._config.bit_count != other._config.bit_count
            or self._config.hash_count != other._config.hash_count
            or self._domain != other._domain
            or self._algorithm != other._algorithm
        ):
            raise IncompatibleMergeError("metadata differs")
        if self._inserted_count > MAX_UINT64 - other._inserted_count:
            raise CountOverflowError("inserted count")
        self._bitset = bytearray(
            left | right
            for left, right in zip(self._bitset, other._bitset, strict=True)
        )
        self._inserted_count += other._inserted_count

    def false_positive_estimate(self) -> float:
        """Return the fill-ratio Bloom false-positive estimate."""

        fill_ratio = float(self.set_bit_count()) / float(self._config.bit_count)
        return math.pow(fill_ratio, float(self._config.hash_count))

    def inserted_count(self) -> int:
        """Return the number of AddHash calls observed."""

        return self._inserted_count

    def bit_count(self) -> int:
        """Return the configured bit count."""

        return self._config.bit_count

    def hash_count(self) -> int:
        """Return the configured hash count."""

        return self._config.hash_count

    def rated_insertions(self) -> int:
        """Return the profile's rated insertion count."""

        return self._config.rated_insertions

    def target_fpr(self) -> float:
        """Return the profile's target false-positive rate."""

        return self._config.target_fpr

    def set_bit_count(self) -> int:
        """Return the number of one bits in the serialized bitset."""

        return sum(byte.bit_count() for byte in self._bitset)

    def bitset(self) -> bytes:
        """Return the canonical bitset bytes."""

        return bytes(self._bitset)

    def set_bits(self) -> list[int]:
        """Return set bit positions in ascending order."""

        out: list[int] = []
        for byte_index, byte in enumerate(self._bitset):
            for bit in range(8):
                if byte & (1 << bit):
                    out.append(byte_index * 8 + bit)
        return out

    def clone(self) -> Sketch:
        """Return a deep copy of the sketch."""

        clone = type(self)(self._profile, self._domain, self._algorithm)
        clone._bitset = bytearray(self._bitset)
        clone._inserted_count = self._inserted_count
        return clone

    def marshal_binary(self) -> bytes:
        """Return deterministic protobuf bytes."""

        return _proto.serialize(self._to_proto())

    def _location(self, value: int, index: int) -> int:
        return hashfamily.bloom_hash(value, index) % self._config.bit_count

    def _set_bit(self, position: int) -> None:
        self._bitset[position // 8] |= 1 << (position % 8)

    def _bit(self, position: int) -> bool:
        return self._bitset[position // 8] & (1 << (position % 8)) != 0

    def _to_proto(self) -> Any:
        metadata = _proto.sketchpb.SketchMetadata(
            kind=_proto.SKETCH_KIND_BLOOM,
            wire_version=_proto.WIRE_VERSION,
            profile=self._profile,
            hash_domain=self._domain,
            hash_algo=_proto.HASH_ALGORITHM_HMAC_SHA256_64,
            representation_mode=_proto.REPRESENTATION_BLOOM_BITSET,
        )
        metadata.bloom_bit_count = self._config.bit_count
        metadata.bloom_hash_count = self._config.hash_count

        body = _proto.sketchpb.BloomSketch(
            bitset=bytes(self._bitset),
            inserted_count=self._inserted_count,
        )
        return _proto.sketchpb.Sketch(metadata=metadata, bloom=body)


def new(
    profile: str,
    domain: str = profiles.PROMPT_V1,
    algorithm: str = profiles.HMAC_SHA256_64,
) -> Sketch:
    """Construct an empty Bloom sketch."""

    return Sketch(profile, domain, algorithm)


def parse(data: bytes) -> Sketch:
    """Decode a Bloom sketch from deterministic protobuf bytes."""

    return Sketch.parse(data)


def _byte_len(bit_count: int) -> int:
    return (bit_count + 7) // 8


def _final_byte_padding_zero(bitset: bytes, bit_count: int) -> bool:
    remainder = bit_count % 8
    if remainder == 0 or not bitset:
        return True
    mask = 0xFF << remainder
    return bitset[-1] & mask == 0
